### PR TITLE
Feature/terms and conditions

### PR DESCRIPTION
**Pull request by ronaldmansveld**
Merge all FE and node-proxy layer work for TC

### CODE REVIEW

File: api/termsAndConditions/index.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+export async function getTermsAndConditions(req: express.Request, res: express.Response) {
+    let errReport: any
+    try {
+        const url = getTermsAndConditionsUrl(config.services.termsAndConditions, application.idamClient)
+        const response = await http.get(url)
+        res.send(response.data)
+    } catch (error) {
+        // we get a 404 if the user has not agreed to Terms and conditions
+        errReport = {
+            apiError: error.data.message,
```
Comment (fazanki) [inline api/termsAndConditions/index.ts]: this error should be managed through common error method

File: api/userTermsAndConditions/index.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import { GetUserAcceptTandCResponse, PostUserAcceptTandCResponse } from '../interface/userAcceptTandCResponse'
+import { http } from '../lib/http'
+import { isUserTandCPostSuccessful } from '../lib/util'
+import { getUserTermsAndConditionsUrl, postUserTermsAndConditionsUrl } from './userTermsAndConditionsUtil'
+
+export async function getUserTermsAndConditions(req: express.Request, res: express.Response) {
+    let errReport: any
+    if (!req.params.userId) {
+        errReport = {
+            apiError: 'UserId is missing',
```
Comment (fazanki) [inline api/userTermsAndConditions/index.ts]: this error should be managed through common error method

File: api/userTermsAndConditions/index.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    }
+    try {
+        const url = getUserTermsAndConditionsUrl(config.services.termsAndConditions, req.params.userId, application.idamClient)
+        const response = await http.get(url)
+        const userTandCResponse = response.data as GetUserAcceptTandCResponse
+        res.send(userTandCResponse.accepted)
+    } catch (error) {
+        // we get a 404 if the user has not agreed to Terms and conditions
+        if (error.status === 404) {
+            res.send(false)
```
Comment (fazanki) [inline api/userTermsAndConditions/index.ts]: for readability improvements, we should have a neum/constant file that stores all the server response codes.

File: api/userTermsAndConditions/index.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            return
+        }
+        errReport = {
+            apiError: error.data.message,
```
Comment (fazanki) [inline api/userTermsAndConditions/index.ts]: common error handling is needed

File: api/userTermsAndConditions/index.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        }
+        res.status(error.status).send(errReport)
+    }
+}
+
+export async function postUserTermsAndConditions(req: express.Request, res: express.Response) {
+    let errReport: any
+    if (!req.body.userId) {
+        errReport = {
+            apiError: 'UserId is missing',
```
Comment (fazanki) [inline api/userTermsAndConditions/index.ts]: common error handling

File: api/userTermsAndConditions/index.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    }
+    try {
+        const data = {userId: req.body.userId}
+        const url = postUserTermsAndConditionsUrl(config.services.termsAndConditions, application.idamClient)
+        const response = await http.post(url, data)
+        const postResponse = response.data as PostUserAcceptTandCResponse
+        res.send(isUserTandCPostSuccessful(postResponse, req.body.userId))
+    } catch (error) {
+        errReport = {
+            apiError: error.data.message,
```
Comment (fazanki) [inline api/userTermsAndConditions/index.ts]: common error handling

File: src/app/accept-tc.wrapper/accept-tc-wrapper.component.html [hunk 7]
```diff
@@ -0,0 +1,6 @@
+<exui-app-header></exui-app-header>
+<exui-accept-tc-component
+  [uId]="uId$ | async"
+  (acceptTC)="onAcceptTandC()"
+></exui-accept-tc-component>
+<exui-app-footer></exui-app-footer>
```
Comment (fazanki) [inline src/app/accept-tc.wrapper/accept-tc-wrapper.component.html]: not sure how this passed linting we need an extra line at the end of the file

File: src/app/containers/terms-and-conditions/terms-and-conditions.component.html [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+      <ng-container *ngIf="document; else loading">
+        <xuilib-terms-and-conditions [document]="document"></xuilib-terms-and-conditions>
+      </ng-container>
+      <ng-template #loading>Loading...</ng-template>
+
+    </div>
+  </div>
+  </main>
+</div>
+<exui-app-footer></exui-app-footer>
```
Comment (fazanki) [inline src/app/containers/terms-and-conditions/terms-and-conditions.component.html]: missing extra line

File: src/app/containers/terms-and-conditions/terms-and-conditions.component.ts [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+export class TermsAndConditionsComponent implements OnInit {
+
+    public document: TCDocument = null;
+
+    constructor(private readonly store: Store<fromRoot.State>) {
+    }
+
+    public ngOnInit() {
+        this.store.pipe(
+            select(fromRoot.getTermsAndConditions)
```
Comment (fazanki) [inline src/app/containers/terms-and-conditions/terms-and-conditions.component.ts]: it needs to complete memory leeks

File: api/termsAndConditions/index.ts (hunk 1 above)
Comment (ronaldmansveld) [reply api/termsAndConditions/index.ts]: Most errors are different, so no point in handling them the same. Also, different endpoints will want to return different errors based on different conditions.

File: api/userTermsAndConditions/index.ts (hunk 2 above)
Comment (ronaldmansveld) [reply api/userTermsAndConditions/index.ts]: Most errors are different, so no point in handling them the same. Also, different endpoints will want to return different errors based on different conditions.

File: api/userTermsAndConditions/index.ts (hunk 3 above)
Comment (ronaldmansveld) [reply api/userTermsAndConditions/index.ts]: A constant file for http status codes? I feel that's a bit over the top. I agree that we shouldn't use magic numbers, but http status codes are not magic numbers.

File: api/userTermsAndConditions/index.ts (hunk 4 above)
Comment (ronaldmansveld) [reply api/userTermsAndConditions/index.ts]: Most errors are different, so no point in handling them the same. Also, different endpoints will want to return different errors based on different conditions.

File: api/userTermsAndConditions/index.ts (hunk 5 above)
Comment (ronaldmansveld) [reply api/userTermsAndConditions/index.ts]: Most errors are different, so no point in handling them the same. Also, different endpoints will want to return different errors based on different conditions.

File: api/userTermsAndConditions/index.ts (hunk 6 above)
Comment (ronaldmansveld) [reply api/userTermsAndConditions/index.ts]: Most errors are different, so no point in handling them the same. Also, different endpoints will want to return different errors based on different conditions.

File: src/app/containers/terms-and-conditions/terms-and-conditions.component.ts (hunk 9 above)
Comment (ronaldmansveld) [reply src/app/containers/terms-and-conditions/terms-and-conditions.component.ts]: If you mean that we should save a reference to the subscription to unsubscribe in an ngDestroy, then yes, I agree.
If you mean something else, please comment further.

File: src/app/guards/acceptTerms.guard.ts [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  canActivate(): Observable<boolean> {
+    const isPuiCaseManager = AppUtils.isRoleExistsForUser('pui-case-manager', this.cookieService);
+    if (!isPuiCaseManager) {
+      return of(true);
+    }
+    return this.store.pipe(
+      select(fromApp.getTandCLoaded),
+      tap(tc => {
+        if (!tc.isLoaded) {
+          const userId = this.cookieService.get('__userid__');
```
Comment (paddybasi) [inline src/app/guards/acceptTerms.guard.ts]: please get this value from the envService

**Pull request merged**